include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp

```diff
@@ -185,86 +185,43 @@ namespace alpaka
         template<typename TApi, bool TBlocking, typename TTask>
         struct Enqueue<uniform_cuda_hip::detail::QueueUniformCudaHipRt<TApi, TBlocking>, TTask>
         {
-            enum class CallbackState
-            {
-                enqueued,
-                notified,
-                finished,
-            };
-
-            struct CallbackSynchronizationData : public std::enable_shared_from_this<CallbackSynchronizationData>
-            {
-                std::mutex m_mutex;
-                std::condition_variable m_event;
-                CallbackState m_state = CallbackState::enqueued;
+            using QueueImpl = uniform_cuda_hip::detail::QueueUniformCudaHipRtImpl<TApi>;
+
+            struct HostFuncData
+            {
+                QueueImpl& q; // We don't need to keep the queue alive, because in it's dtor it will synchronize with
+                              // the CUDA/HIP stream and wait until all host functions and the CallbackThread are done.
+                              // It's actually an error ot copy the queue into the host function. Destroying it here
+                              // would call CUDA/HIP APIs from the host function. Passing it further to the Callback
```

Comment:
> Destroying it here would call CUDA/HIP APIs from the host function.

Why that would be a problem ?

Reply:
Because the CUDA documentation for `cudaLaunchHostFunc` [says](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1g05841eaa5f90f27124241baafb3e856f):

> The host function must not make any CUDA API calls. Attempting to use a CUDA API may result in cudaErrorNotPermitted but this is not required.

Reply:
Could you please move the comment above the expression? This will make the lines more readable